### PR TITLE
fix: support cors multi domain and wildcard in ingestion server

### PR DESCRIPTION
**Pull request by dengmingtong**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary
support cors multi domain and wildcard in ingestion server

## Implementation highlights

Change nginx conf to support cors multi domain and wildcard in ingestion server.
The following is the table of UI input, supported cors domains and ingestion stack input parameter(ServerCorsOrigin)

UI Input | Valid domains | Input stack string
-- | -- | --
https://example.com | https://example.com | "https://example\.com“
http://example.com | http://example.com | "http://example\.com“
http://\*example\*.com | http://xxxexamplexxx.com, http://example.com | "http://.\*example.\*\\.com“
https://\*example\*.com | https://xxxexamplexxx.com, https://example.com | "https://.\*example.\*\\.com“
http://example.com,https://abc.com | http://example.com, https://abc.com | "http://example\.com\|https://abc\.com"
http://localhost | http://localhost:xxxx | "http://localhost“
http://127.0.0.1 | http://127.0.0.1:xxxx | "http://127.0.0.1“
|* | Any domain | ".*"

 

## Test checklist

- [ ] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [x] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [x] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

**Comment by zxkane** [comment]
Let's merge with https://github.com/awslabs/clickstream-analytics-on-aws/pull/158

File: src/ingestion-server/server/images/nginx/config/nginx.conf [hunk 1]
```diff
@@ -40,10 +40,12 @@ http {
             # limit_req_status 444;
 
             add_header Content-Type text/plain;
-            add_header 'Access-Control-Allow-Origin' '%%SERVER_CORS_ORIGIN%%';
-            add_header 'Access-Control-Allow-Credentials' 'true';
-            add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
-            add_header 'Access-Control-Allow-Headers' 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type';
+            if ($http_origin ~* (%%SERVER_CORS_ORIGIN%%)) {
```
Comment (zxkane) [inline src/ingestion-server/server/images/nginx/config/nginx.conf:43]: Will it work if the `SERVER_CORS_ORIGIN` is blank or empty?

Comment (zxkane) [reply src/ingestion-server/server/images/nginx/config/nginx.conf:43]: How about inputting `http://localhost:8080`? Will only allow the requests from localhost with port 8080?

Comment (dengmingtong) [reply src/ingestion-server/server/images/nginx/config/nginx.conf:43]: Good point,  
After refactoring, if SERVER_CORS_ORIGIN is blank or empty, no domain will be allowed .

For localhost, if input http://localhost:8080, will only allow http://localhost:8080/xxx, other port will not allow.

**Comment by dengmingtong** [comment]
> Let's merge with #158

Done

**Comment by zxkane** [comment]
duplicated with #158

**Pull request closed without merging**